src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs

```diff
@@ -0,0 +1,146 @@
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Caches the semantic token information that needs to be preserved between multiple
+    /// semantic token requests.
+    /// Multiple token sets can be cached per document. The number of token sets cached
+    /// per document is determined by the _maxCachesPerDoc field.
+    /// </summary>
+    [Export(typeof(SemanticTokensCache)), Shared]
+    internal class SemanticTokensCache
+    {
+        /// <summary>
+        /// Maps a LSP token type to its respective index recognized by LSP.
+        /// </summary>
+        public static readonly Dictionary<string, int> TokenTypesToIndex;
+
+        /// <summary>
+        /// Number of cached token sets we store per document. Must be >= 1.
+        /// </summary>
+        private readonly int _maxCachesPerDoc = 5;
+
+        /// <summary>
+        /// The next resultId available to use.
+        /// </summary>
+        private long _nextResultId;
+
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the _tokens dictionary.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        #region protected by _semaphore
+        /// <summary>
+        /// Maps a document URI to its n most recently cached token sets.
+        /// </summary>
+        private readonly Dictionary<Uri, List<LSP.SemanticTokens>> _tokens =
+            new Dictionary<Uri, List<LSP.SemanticTokens>>();
+        #endregion
+
+        static SemanticTokensCache()
+        {
+            // Computes the mapping between a LSP token type and its respective index recognized by LSP.
+            TokenTypesToIndex = new Dictionary<string, int>();
+            for (var i = 0; i < LSP.SemanticTokenTypes.AllTypes.Count; i++)
+            {
+                TokenTypesToIndex.Add(LSP.SemanticTokenTypes.AllTypes[i], i);
+            }
+        }
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SemanticTokensCache()
+        {
+        }
+
+        /// <summary>
+        /// Updates the given document's token set cache. Removes old cache results if the document's
+        /// cache is full.
+        /// </summary>
+        public async Task UpdateCacheAsync(
+            Uri uri,
+            LSP.SemanticTokens tokens,
+            CancellationToken cancellationToken)
+        {
+            // If the resultId of the semantic tokens is null, don't cache anything since we'll
+            // be unable to retrieve the results later.
+            if (tokens.ResultId == null)
```

Comment:
should this be a contract check instead.  i.e. who is the caller of this that is passing in tokens that dont' have a result id?